File: src/main/java/com/github/sorusclient/client/bootstrap/BootstrapManager.java
/*
 * Copyright (c) 2022. Danterus
 * Copyright (c) 2022. Sorus Contributors
 *
 * SPDX-License-Identifier: MPL-2.0
 */

package com.github.sorusclient.client.bootstrap;

import com.github.sorusclient.client.transform.Transformer;
import lombok.SneakyThrows;
import org.apache.commons.io.IOUtils;
import org.json.JSONObject;

import java.lang.reflect.InvocationTargetException;
import java.net.URL;
import java.nio.charset.StandardCharsets;

public class BootstrapManager {

    @SneakyThrows
    public static JSONObject loadJson(URL path, String minecraftVersion) {
        var rawString = IOUtils.toString(path, StandardCharsets.UTF_8);
        var json = new JSONObject(rawString);

        if (json.has("versions")) {
            for (var entry : json.getJSONObject("versions").toMap().entrySet()) {
                if (entry.getKey().equals(minecraftVersion)) {
                    loadJson(BootstrapManager.class.getClassLoader().getResource((String) entry.getValue()), minecraftVersion);
                }
            }
        }

        var namespace = json.has("namespace") ? json.getString("namespace") : null;

        if (json.has("transformers")) {
            for (var transformer : json.getJSONArray("transformers")) {
                var transformer2 = (Class<? extends Transformer>) Class.forName((namespace != null ? namespace + "." : "") + transformer);
                addTransformer(transformer2);
            }
        }

        if (json.has("initializers")) {
            for (var initializer : json.getJSONArray("initializers")) {
                var initializer2 = ((Class<? extends Initializer>) Class.forName((namespace != null ? namespace + "." : "") + initializer)).getConstructor().newInstance();
                initializer2.initialize();
            }
        }

        return json;
    }

    public static void addTransformer(Class<? extends Transformer> transformer) {
        var classLoader = BootstrapManager.class.getClassLoader();
        try {
            classLoader.getClass().getMethod("addTransformer", Class.class).invoke(classLoader, transformer);
        } catch (IllegalAccessException | InvocationTargetException | NoSuchMethodException e) {
            throw new IllegalStateException(e);
        }
    }

    public static void addURL(URL url) {
        var classLoader = BootstrapManager.class.getClassLoader();
        try {
            classLoader.getClass().getMethod("addURL", Class.class).invoke(classLoader, url);
        } catch (IllegalAccessException | InvocationTargetException | NoSuchMethodException e) {
            throw new IllegalStateException(e);
        }
    }

    public static void removeURL(URL url) {
        var classLoader = BootstrapManager.class.getClassLoader();
        try {
            classLoader.getClass().getMethod("removeURL", Class.class).invoke(classLoader, url);
        } catch (IllegalAccessException | InvocationTargetException | NoSuchMethodException e) {
            throw new IllegalStateException(e);
        }
    }

}

File: src/main/kotlin/com/github/sorusclient/client/bootstrap/BootstrapManager.kt
package com.github.sorusclient.client.bootstrap

import com.github.sorusclient.client.bootstrap.transformer.Transformer
import org.apache.commons.io.IOUtils
import org.json.JSONObject
import java.net.URL
import java.nio.charset.StandardCharsets

object BootstrapManager {

    fun loadJson(path: URL, minecraftVersion: String): JSONObject {
        val json = run {
            val rawString = IOUtils.toString(path, StandardCharsets.UTF_8)
            JSONObject(rawString)
        }

        if (json.has("versions")) {
            for ((key, value) in json.getJSONObject("versions").toMap()) {
                if (key == minecraftVersion) {
                    loadJson(BootstrapManager::class.java.classLoader.getResource(value as String)!!, minecraftVersion)
                }
            }
        }

        val namespace = if (json.has("namespace")) { json.getString("namespace") } else { null }

        if (json.has("transformers")) {
            for (transformer in json.getJSONArray("transformers")) {
                val transformer = Class.forName(if (namespace != null) { "$namespace." } else { "" } + transformer as String) as Class<out Transformer>
                addTransformer(transformer)
            }
        }

        if (json.has("initializers")) {
            for (initializer in json.getJSONArray("initializers")) {
                val initializer = Class.forName(if (namespace != null) { "$namespace." } else { "" } + initializer as String).newInstance() as Initializer
                initializer.initialize()
            }
        }

        return json
    }

    fun addTransformer(transformer: Class<out Transformer>) {
        val classLoader = BootstrapManager::class.java.classLoader
        classLoader.javaClass.getMethod("addTransformer", Class::class.java).invoke(classLoader, transformer)
    }

    fun addURL(url: URL) {
        val classLoader = BootstrapManager::class.java.classLoader
        classLoader.javaClass.getMethod("addURL", Class::class.java).invoke(classLoader, url)
    }

    fun removeURL(url: URL) {
        val classLoader = BootstrapManager::class.java.classLoader
        classLoader.javaClass.getMethod("removeURL", Class::class.java).invoke(classLoader, url)
    }

}